lib/ex_webrtc/peer_connection.ex

```diff
@@ -470,91 +458,138 @@ defmodule ExWebRTC.PeerConnection do
-    state = set_description(:local, type, sdp, state)
+  defp apply_local_description(%SessionDescription{type: :rollback}, _state) do
+    {:error, :not_implemented}
+  end
 
-    demuxer = %Demuxer{
-      state.demuxer
-      | extensions: SDPUtils.get_extensions(sdp),
-        pt_to_mid: SDPUtils.get_payload_types(sdp)
-    }
+  defp apply_local_description(%SessionDescription{type: type, sdp: raw_sdp}, state) do
+    with {:ok, next_sig_state} <- next_state(state.signaling_state, :local, type),
+         :ok <- check_altered(type, raw_sdp, state),
+         {:ok, sdp} <- parse_sdp(raw_sdp) do
+      mid_ext_id =
+        sdp
+        |> SDPUtils.get_extensions()
+        |> Enum.find_value(fn
+          {id, {ExRTP.Packet.Extension.SourceDescription, :mid}} -> id
+          _ -> nil
+        end)
 
-    if type == :answer do
-      {:setup, setup} = ExSDP.Media.get_attribute(hd(sdp.media), :setup)
-      DTLSTransport.start_dtls(state.dtls_transport, setup, state.peer_fingerprint)
-    end
+      pt_to_mid = SDPUtils.get_payload_to_mid(sdp)
+      demuxer = %Demuxer{state.demuxer | mid_ext_id: mid_ext_id, pt_to_mid: pt_to_mid}
+      state = %{state | demuxer: demuxer}
 
-    {:ok, %{state | transceivers: new_transceivers, demuxer: demuxer}}
-  end
+      if state.ice_gathering_state == :new do
+        state.ice_transport.gather_candidates(state.ice_pid)
+      end
+
+      # TODO: do rest of the stuff from w3c
 
-  defp update_local_transceivers(:offer, transceivers, _sdp) do
-    # TODO: at the moment, we assign mids in create_offer
-    transceivers
+      state = set_description(:local, type, sdp, state)
+      {:ok, %{state | signaling_state: next_sig_state}}
+    end
   end
 
-  defp update_local_transceivers(:answer, transceivers, _sdp) do
-    transceivers
+  defp apply_remote_description(%SessionDescription{type: :rollback}, _state) do
+    {:error, :not_implemented}
   end
 
-  defp apply_remote_description(type, sdp, state) do
-    # TODO apply steps listed in RFC 8829 5.10
-    with :ok <- SDPUtils.ensure_mid(sdp),
+  defp apply_remote_description(%SessionDescription{type: type, sdp: raw_sdp}, state) do
+    with {:ok, next_sig_state} <- next_state(state.signaling_state, :remote, type),
+         {:ok, sdp} <- parse_sdp(raw_sdp),
+         :ok <- SDPUtils.ensure_mid(sdp),
          :ok <- SDPUtils.ensure_bundle(sdp),
          {:ok, {ice_ufrag, ice_pwd}} <- SDPUtils.get_ice_credentials(sdp),
          {:ok, {:fingerprint, {:sha256, peer_fingerprint}}} <- SDPUtils.get_cert_fingerprint(sdp),
-         {:ok, new_transceivers} <-
-           update_remote_transceivers(state.transceivers, sdp, state.config) do
-      :ok =
-        state.ice_transport.set_remote_credentials(state.ice_pid, ice_ufrag, ice_pwd)
+         {:ok, dtls_role} <- SDPUtils.dtls_role_from_remote(sdp),
+         {:ok, transceivers} <- update_transceivers(sdp, state) do
+      state = %{state | can_trickle_ice_candidates: SDPUtils.check_if_trickle(sdp)}
```

Comment:
Avoid adding fields we are currently not using. We will add them once they are needed